### ISSUE TITLE
Introduce new string type for names/keys with length limit

### ISSUE THREAD

**Issue by inexorabletash** [heycam/webidl#552]
Many places in the platform allow DOMStrings for map keys or resource names, e.g. localStorage keys, Indexed DB database/store/index names, Service Worker Cache names, and so on. Specs do not define length limits, making it up to the UA to impose a limit to prevent OOMs. Rather than forcing each spec to specify a limit and behavior, we could introduce a new string type in WebIDL so that specs just need to use the type and bindings code in each UA will do the work.

Context:
* https://github.com/w3c/ServiceWorker/issues/1310 (SW's Cache API)
* https://github.com/inexorabletash/web-locks/issues/36 (proposal for a Locks API)
* https://github.com/w3c/IndexedDB/issues/207 (IDB discussion about problems with arbitrary length strings; we didn't do anything specific there)

Plausible names include "KeyString" or "NameString", but I'm not a huge fan of either. "Key" only makes sense if the map is fairly explicit. "Name" could imply that these are human readable and thus should be normalized, have locale-aware collation, etc.

Since I'm a nerd I have a hard time not picking a power-of-two limit (e.g. 1024), or feeling like I'm intentionally avoiding picking such a number (e.g. 1000).

But as a straw man:

> **NameString**
> The NameString type corresponds to the set of all possible sequences of code units up to length 1000. Such sequences are commonly interpreted as UTF-16 encoded strings although this is not required. 
>
>> Specifications should use the NameString type to represent strings that are provided by script to serve as the programmatic names of resources, such as the names of database tables or keys in a configuration map. These strings are assumed to have a fixed length within a web application, and to not contain arbitrary user data.

And the ECMAScript binding steps would be something like:

> An ECMAScript value V is converted to an IDL NameString value by running the following algorithm:
> 1. Let x be ToString(V).
> 1. If the length of x is greater than 1000, then throw a TypeError.
> 1. Return the IDL NameString value that represents the same sequence of code units as the one the ECMAScript String value x represents.
> 
> The result of converting an IDL NameString value to an ECMAScript value is the String value that represents the same sequence of code units that the IDL NameString represents.

See also: 
* #335 - i.e. we'd probably want something over in [Infra](https://infra.spec.whatwg.org/) too

**Comment by domenic** [comment]
Bikeshed: LimitedLengthString

**Comment by Zirro** [comment]
If this is intended for use with existing APIs - particularly ones like localStorage which have been around for some time - there might be web compatibility concerns. I could imagine certain edge cases/misuses out there pushing names above 1000 characters.

To be on the safe side, it's probably best to investigate which limits browsers currently use for the relevant APIs - and potentially add a use counter to detect unusually long names as well.

**Comment by mkruisselbrink** [comment]
Yeah, agreed that we'd need to be careful introducing limits that are orders of magnitude smaller than what implementations currently happen to allow. But I do think it's worth trying to have such explicit limits for as many APIs as possible.

**Comment by annevk** [comment]
Do all of the potential candidates want DOMString?

**Comment by inexorabletash** [comment]
> Do all of the potential candidates want DOMString?

As opposed to USVString or ByteString? So far, yes. We've definitely had implementation bugs in various APIs where UTF-8 conversion was done internally, but I'm not aware of any cases where we flipped a spec to USVString.

**Comment by domenic** [comment]
We could also use an extended attribute for this these days, now that they're attached to types. That would allow it to work with USVString or ByteString as well.

**Comment by inexorabletash** [comment]
For the record, I assume the path forward here is for a willing engine to annotate potential strings in the API with a non-standard extended attribute and gather metrics on the lengths. With that data, we'll have confidence about the compatibility of doing this.

We should compile a list of targets. I suppose all *Strings are candidates. Only 1600 or so methods/dict entries/writable attributes found in Blink to inspect - wheeee!

**Comment by annevk** [comment]
@inexorabletash I don't think it has to be a success for all of them for this to be a success, for what it's worth. And I can also see varying the limit.

**Comment by tobie** [comment]
> And I can also see varying the limit.

Yeah, there are multiple options:

```webidl
[MaxLength=1024] DOMString;
[MaxLength=short] DOMString;
// etc.
```
